pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -0,0 +1,465 @@
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				// APIResource.Kind is not used by the dynamic client, so
+				// leave it empty. We want to list this resource in all
+				// namespaces if it's namespace scoped, so leave
+				// APIResource.Namespaced as false is all right.
+				apiResource := unversioned.APIResource{Name: resource.Resource}
+				return client.Resource(&apiResource, api.NamespaceAll).Watch(&options)
+			},
+		},
+		nil,
+		ResourceResyncTime,
+		framework.ResourceEventHandlerFuncs{
+			// add the event to the propagator's eventQueue.
+			AddFunc: func(obj interface{}) {
+				event := event{
+					eventType: addEvent,
+					obj:       obj,
+				}
+				p.eventQueue.Add(event)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				event := event{updateEvent, newObj, oldObj}
+				p.eventQueue.Add(event)
+			},
+			DeleteFunc: func(obj interface{}) {
+				event := event{
+					eventType: deleteEvent,
+					obj:       obj,
+				}
+				p.eventQueue.Add(event)
+			},
+		},
+	)
+	return monitor, nil
+}
+
+var ignoredResources = map[unversioned.GroupVersionResource]struct{}{
+	unversioned.GroupVersionResource{Group: "extensions", Version: "v1beta1", Resource: "replicationcontrollers"}: {},
+	unversioned.GroupVersionResource{Group: "", Version: "v1", Resource: "bindings"}:                              {},
+	unversioned.GroupVersionResource{Group: "", Version: "v1", Resource: "componentstatuses"}:                     {},
+	unversioned.GroupVersionResource{Group: "", Version: "v1", Resource: "events"}:                                {},
+}
+
+func NewGarbageCollector(clientPool dynamic.ClientPool, resources []unversioned.GroupVersionResource) (*GarbageCollector, error) {
+	gc := &GarbageCollector{
+		clientPool: clientPool,
+		dirtyQueue: workqueue.New(),
+		// TODO: should use a dynamic RESTMapper built from the discovery results.
+		restMapper: registered.RESTMapper(),
+	}
+	gc.propagator = &Propagator{
+		eventQueue: workqueue.New(),
+		uidToNode:  make(map[types.UID]*node),
+		gc:         gc,
+	}
+	for _, resource := range resources {
+		if _, ok := ignoredResources[resource]; ok {
+			glog.V(6).Infof("ignore resource %#v", resource)
+			continue
+		}
+		monitor, err := monitorFor(gc.propagator, gc.clientPool, resource)
+		if err != nil {
+			return nil, err
+		}
+		gc.monitors = append(gc.monitors, monitor)
+	}
+	return gc, nil
+}
+
+func (gc *GarbageCollector) worker() {
+	key, quit := gc.dirtyQueue.Get()
+	if quit {
+		return
+	}
+	defer gc.dirtyQueue.Done(key)
+	err := gc.processItem(key.(*node))
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Error syncing item %v: %v", key, err))
+	}
+}
+
+// apiResource consults the REST mapper to translate an <apiVersion, kind,
+// namespace> tuple to a unversioned.APIResource struct.
+func (gc *GarbageCollector) apiResource(apiVersion, kind, namespace string) (*unversioned.APIResource, error) {
```

Comment:
Consider changing it to take a `namespaced bool`. I think that would be clearer?